File: python/src/workshop_api/fitness/managing_plans/create_plan/create_plan_controller.py
from uuid import uuid4

from fastapi import APIRouter, Depends, Response, status
from sqlalchemy.orm import Session

from workshop_api.fitness.database import get_db_session
from workshop_api.fitness.managing_plans.shared.plan_entity import PlanOrmModel
from workshop_api.fitness.managing_plans.shared.plan_response import PlanResponse
from workshop_api.fitness.managing_plans.shared.plan_upsert_request import PlanUpsertRequest

router = APIRouter(prefix="/api/plans", tags=["plan"])


@router.post(
    "",
    response_model=PlanResponse,
    response_model_by_alias=True,
    status_code=status.HTTP_201_CREATED,
)
def create_plan(
    request: PlanUpsertRequest,
    response: Response,
    session: Session = Depends(get_db_session),
) -> PlanResponse:
    plan = PlanOrmModel(
        id=str(uuid4()),
        title=request.title,
        description=request.description,
        duration_in_months=request.duration_in_months,
        price=request.price,
    )

    session.add(plan)
    session.commit()
    session.refresh(plan)

    created = PlanResponse.from_entity(plan)
    response.headers["Location"] = f"/api/plans/{created.id}"
    return created

File: java/src/main/java/com/workshop/architecture/fitness/create_plan/CreatePlanController.java
package com.workshop.architecture.fitness.create_plan;

import com.workshop.architecture.fitness.shared.PlanEntity;
import com.workshop.architecture.fitness.shared.PlanRepository;
import com.workshop.architecture.fitness.shared.PlanResponse;
import com.workshop.architecture.fitness.shared.PlanUpsertRequest;
import jakarta.validation.Valid;
import java.net.URI;
import java.util.UUID;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/plans")
public class CreatePlanController {

    private final PlanRepository planRepository;

    public CreatePlanController(PlanRepository planRepository) {
        this.planRepository = planRepository;
    }

    @PostMapping
    ResponseEntity<PlanResponse> createPlan(@Valid @RequestBody PlanUpsertRequest request) {
        PlanEntity plan = new PlanEntity(
                UUID.randomUUID(),
                request.title(),
                request.description(),
                request.durationInMonths(),
                request.price()
        );

        PlanResponse response = PlanResponse.fromEntity(planRepository.save(plan));
        return ResponseEntity.created(URI.create("/api/plans/" + response.id())).body(response);
    }
}
